src/rp2_common/hardware_pio/include/hardware/pio.h

```diff
@@ -660,10 +668,151 @@ static inline void pio_clkdiv_restart_sm_mask(PIO pio, uint32_t mask) {
+ */
+static inline void pio_set_irq0_source_mask_enabled(PIO pio, uint32_t source_mask, bool enabled) {
+    check_pio_param(pio);
+    invalid_params_if(PIO, source_mask > PIO_INTR_BITS);
+    if (enabled) {
+        hw_set_bits(&pio->inte0, source_mask);
+    } else {
+        hw_clear_bits(&pio->inte0, source_mask);
+    }
+}
+
+/*! \brief  Enable/Disable multiple sources on a PIO's IRQ 1
+ *  \ingroup hardware_pio
+ *
+ * \param source_mask Mask of bits, one for each source number (see \ref pio_interrupt_source) to affect
+ * \param enabled true to enable all the source specified in the mask on IRQ 1, false to disable all the source specified in the mask on IRQ 1
+ */
+static inline void pio_set_irq1_source_mask_enabled(PIO pio, uint32_t source_mask, bool enabled) {
+    check_pio_param(pio);
+    invalid_params_if(PIO, source_mask > PIO_INTR_BITS);
+    if (enabled) {
+        hw_set_bits(&pio->inte1, source_mask);
+    } else {
+        hw_clear_bits(&pio->inte1, source_mask);
+    }
+}
+
+/*! \brief  Determine if a particular source is the cause of IRQ0 on a PIO
+ *  \ingroup hardware_pio
+ *
+ * \param pio The PIO instance; either \ref pio0 or \ref pio1
+ * \param source the source number (see \ref pio_interrupt_source)
+ * \return true if the source a cause of an IRQ, false otherwise
+ */
+static inline bool pio_get_irq0_source_status(PIO pio, enum pio_interrupt_source source) {
+    check_pio_param(pio);
+    invalid_params_if(PIO, source >= 12);
+    return pio->inte0 & (1u << source);
+}
+
+/*! \brief  Determine if a particular source is the cause of IRQ1 on a PIO
+ *  \ingroup hardware_pio
+ *
+ * \param pio The PIO instance; either \ref pio0 or \ref pio1
+ * \param source the source number (see \ref pio_interrupt_source)
+ * \return true if the source a cause of an IRQ, false otherwise
+ */
+static inline bool pio_get_irq1_source_status(PIO pio, enum pio_interrupt_source source) {
+    check_pio_param(pio);
+    invalid_params_if(PIO, source >= 12);
+    return pio->inte1 & (1u << source);
+}
+
+/*! \brief  Acknowledge a particular source causing IRQ0 on a PIO
+ *  \ingroup hardware_pio
+ *
+ * Note that this method only has any affect for pis_irq0 thru ppis_irq3 (for which the underlying interrupt is cleared),
+ * whereas the other other PIO interrupt sources are reflections of the states of the FIFO, so the only way to clear
+ * the status is for the FIFO contents to change appropriately.
+ *
+ * \param pio The PIO instance; either \ref pio0 or \ref pio1
+ * \param source the source number (see \ref pio_interrupt_source)
+ */
+static inline void pio_acknowledge_irq0_source(PIO pio, enum pio_interrupt_source source) {
+    check_pio_param(pio);
+    invalid_params_if(PIO, source != pis_irq0 && source != pis_irq1 && source != pis_irq2 && source != pis_irq3);
+    hw_set_bits(&pio->inte0, (1u << source));
```

Comment:
Looking at the RP2040 datasheet, I wonder if this actually needs to be writing to `irq` rather than `inte0` ?
(of course it's entirely possible that I'm totally misunderstanding this!)